src/aws-cpp-sdk-core/include/aws/core/auth/AWSCredentialsProviderChain.h

```diff
@@ -46,8 +46,10 @@ namespace Aws
             void AddProvider(const std::shared_ptr<AWSCredentialsProvider>& provider) { m_providerChain.push_back(provider); }
 
 
-        private:            
+        private:
             Aws::Vector<std::shared_ptr<AWSCredentialsProvider> > m_providerChain;
+            Aws::Crt::Optional<std::shared_ptr<AWSCredentialsProvider>> m_cachedProvider;
```

Comment:
It is already a shared_ptr, I'm not sure Optional is needed here.
Of course the object is optional, but using `optional` type will create an additional allocation and de-reference

Reply:
good call